'pnpm patch' not working with git urls + commit hash

### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

_No response_

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

- add any git+ssh url to either dev or normal dependencies in package.json
- call 'pnpm patch' on that dependency

### Describe the Bug

We're using an 3rd party dependency via github urls + commit hash like in the following example:

"@company/lib": "git+ssh://git@github.com/company/project#e0f3a5b85f4ee2d3b3ea591ae8e25c7dc7686788"

I wanted to use "pnpm patch" to apply a small temporary patch to that version but unfortunately it does not work because it seems I have to call "pnpm patch @company/lib@version" but there's no version.

The error I'm getting is: " ERR_PNPM_NO_MATCHING_VERSION  No matching version found for xxx"

### Expected Behavior

Dependencies from Git urls can be patched using pnpm patch.

### Which Node.js version are you using?

18.13.

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

Thanks for the reply, but this does not work because the package.json version in the specified commit does not exist yet locally (it was never downloaded I assume) because when I do `pnpm view @company/lib versions` pnpm does not show the version which is used in the remote package.json file (the version "behind" this commit hash. So I cannot apply a patch to it.

If I run `pnpm patch @company/lib` as you suggested, I just get the following error:

 ERR_PNPM_NO_MATCHING_VERSION  No matching version found for @company/lib@1.2.3
The latest release of @company/lib is "0.1.2"
 
 1.2.3 is the version used in the remote package.json file for this commit hash. And 0.1.2 is the latest one available locally as per `pnpm view @company/lib versions`.
 
 

--- a/.changeset/proud-trees-develop.md
+++ b/.changeset/proud-trees-develop.md
@@ -1,0 +1,5 @@
+---
+"@pnpm/plugin-commands-patching": patch
+---
+
+Should fetch dependency from tarball url when patching dependency installed from git [#7196](https://github.com/pnpm/pnpm/issues/7196)

--- a/.changeset/proud-trees-develop.md
+++ b/.changeset/proud-trees-develop.md
@@ -1,5 +1,6 @@
 ---
 "@pnpm/plugin-commands-patching": patch
+"pnpm": patch
 ---
 
 Should fetch dependency from tarball url when patching dependency installed from git [#7196](https://github.com/pnpm/pnpm/issues/7196)

--- a/patching/plugin-commands-patching/package.json
+++ b/patching/plugin-commands-patching/package.json
@@ -53,6 +53,7 @@
     "@pnpm/parse-wanted-dependency": "workspace:*",
     "@pnpm/patching.apply-patch": "workspace:*",
     "@pnpm/pick-registry-for-package": "workspace:*",
+    "@pnpm/pick-fetcher": "workspace:*",
     "@pnpm/plugin-commands-installation": "workspace:*",
     "@pnpm/read-package-json": "workspace:*",
     "@pnpm/read-project-manifest": "workspace:*",

--- a/patching/plugin-commands-patching/src/getPatchedDependency.ts
+++ b/patching/plugin-commands-patching/src/getPatchedDependency.ts
@@ -38,18 +38,18 @@ export async function getPatchedDependency (rawDependency: string, opts: GetPatc
       choices: preferredVersions.map(preferred => ({
         name: preferred.version,
         message: preferred.version,
-        value: preferred.isGitHosted ? preferred.tarball : preferred.version,
-        hint: preferred.isGitHosted ? 'Git Hosted' : undefined,
+        value: preferred.gitTarballUrl ? preferred.gitTarballUrl : preferred.version,
+        hint: preferred.gitTarballUrl ? 'Git Hosted' : undefined,
       })),
       result (selected) {
         const selectedVersion = preferredVersions.find(preferred => preferred.version === selected)!
-        return selectedVersion.isGitHosted ? selectedVersion.tarball : selected
+        return selectedVersion.gitTarballUrl ?? selected
       },
     })
     dep.pref = version
   } else {
     const preferred = preferredVersions[0]
-    dep.pref = preferred.isGitHosted ? preferred.tarball : preferred.version
+    dep.pref = preferred.gitTarballUrl ?? preferred.version
   }
   return dep
 }
@@ -75,8 +75,7 @@ export async function getVersionsFromLockfile (dep: ParseWantedDependencyResult,
       const tarball = (pkgSnapshot.resolution as TarballResolution)?.tarball ?? ''
       return {
         ...nameVerFromPkgSnapshot(depPath, pkgSnapshot),
-        isGitHosted: isGitHostedPkgUrl(tarball ?? ''),
-        tarball,
+        gitTarballUrl: isGitHostedPkgUrl(tarball) ? tarball : undefined,
       }
     })
     .filter(({ name }) => name === pkgName)

--- a/patching/plugin-commands-patching/src/getPatchedDependency.ts
+++ b/patching/plugin-commands-patching/src/getPatchedDependency.ts
@@ -42,7 +42,8 @@ export async function getPatchedDependency (rawDependency: string, opts: GetPatc
         hint: preferred.isGitHosted ? 'Git Hosted' : undefined,
       })),
       result (selected) {
-        return preferredVersions.find(preferred => preferred.version === selected)!.tarball
+        const selectedVersion = preferredVersions.find(preferred => preferred.version === selected)!
+        return selectedVersion.isGitHosted ? selectedVersion.tarball : selected
       },
     })
     dep.pref = version

--- a/patching/plugin-commands-patching/src/getPatchedDependency.ts
+++ b/patching/plugin-commands-patching/src/getPatchedDependency.ts
@@ -38,7 +38,7 @@ export async function getPatchedDependency (rawDependency: string, opts: GetPatc
       choices: preferredVersions.map(preferred => ({
         name: preferred.version,
         message: preferred.version,
-        value: preferred.gitTarballUrl ? preferred.gitTarballUrl : preferred.version,
+        value: preferred.gitTarballUrl ?? preferred.version,
         hint: preferred.gitTarballUrl ? 'Git Hosted' : undefined,
       })),
       result (selected) {

--- a/patching/plugin-commands-patching/src/getPatchedDependency.ts
+++ b/patching/plugin-commands-patching/src/getPatchedDependency.ts
@@ -1,16 +1,17 @@
 import path from 'path'
 import { parseWantedDependency, type ParseWantedDependencyResult } from '@pnpm/parse-wanted-dependency'
 import { prompt } from 'enquirer'
-import { readCurrentLockfile } from '@pnpm/lockfile-file'
+import { readCurrentLockfile, type TarballResolution } from '@pnpm/lockfile-file'
 import { nameVerFromPkgSnapshot } from '@pnpm/lockfile-utils'
 import { PnpmError } from '@pnpm/error'
 import { WANTED_LOCKFILE } from '@pnpm/constants'
 import { readModulesManifest } from '@pnpm/modules-yaml'
+import { isGitHostedPkgUrl } from '@pnpm/pick-fetcher'
 import realpathMissing from 'realpath-missing'
 import semver from 'semver'
 import { type Config } from '@pnpm/config'
 
-type GetPatchedDependencyOptions = {
+export type GetPatchedDependencyOptions = {
   lockfileDir: string
 } & Pick<Config, 'virtualStoreDir' | 'modulesDir'>
 
@@ -22,7 +23,7 @@ export async function getPatchedDependency (rawDependency: string, opts: GetPatc
   if (!preferredVersions.length) {
     throw new PnpmError(
       'PATCH_VERSION_NOT_FOUND',
-      `Can not find ${rawDependency} in project ${opts.lockfileDir}, ${versions.length ? `you can specify currently installed version: ${versions.join(', ')}.` : `did you forget to install ${rawDependency}?`}`
+      `Can not find ${rawDependency} in project ${opts.lockfileDir}, ${versions.length ? `you can specify currently installed version: ${versions.map(({ version }) => version).join(', ')}.` : `did you forget to install ${rawDependency}?`}`
     )
   }
 
@@ -34,17 +35,25 @@ export async function getPatchedDependency (rawDependency: string, opts: GetPatc
       type: 'select',
       name: 'version',
       message: 'Choose which version to patch',
-      choices: versions,
+      choices: preferredVersions.map(preferred => ({
+        name: preferred.version,
+        message: preferred.version,
+        value: preferred.isGitHosted ? preferred.tarball : preferred.version,
+        hint: preferred.isGitHosted ? 'Git Hosted' : undefined,
+      })),
+      result (selected) {
+        return preferredVersions.find(preferred => preferred.version === selected)!.tarball
+      },
     })
     dep.pref = version
   } else {
-    dep.pref = preferredVersions[0]
+    const preferred = preferredVersions[0]
+    dep.pref = preferred.isGitHosted ? preferred.tarball : preferred.version
   }
-
   return dep
 }
 
-async function getVersionsFromLockfile (dep: ParseWantedDependencyResult, opts: GetPatchedDependencyOptions) {
+export async function getVersionsFromLockfile (dep: ParseWantedDependencyResult, opts: GetPatchedDependencyOptions) {
   const modulesDir = await realpathMissing(path.join(opts.lockfileDir, opts.modulesDir ?? 'node_modules'))
   const modules = await readModulesManifest(modulesDir)
   const lockfile = (modules?.virtualStoreDir && await readCurrentLockfile(modules.virtualStoreDir, {
@@ -61,12 +70,18 @@ async function getVersionsFromLockfile (dep: ParseWantedDependencyResult, opts: 
   const pkgName = dep.alias && dep.pref ? dep.alias : (dep.pref ?? dep.alias)
 
   const versions = Object.entries(lockfile.packages ?? {})
-    .map(([depPath, pkgSnapshot]) => nameVerFromPkgSnapshot(depPath, pkgSnapshot))
+    .map(([depPath, pkgSnapshot]) => {
+      const tarball = (pkgSnapshot.resolution as TarballResolution)?.tarball ?? ''
+      return {
+        ...nameVerFromPkgSnapshot(depPath, pkgSnapshot),
+        isGitHosted: isGitHostedPkgUrl(tarball ?? ''),
+        tarball,
+      }
+    })
     .filter(({ name }) => name === pkgName)
-    .map(({ version }) => version)
 
   return {
     versions,
-    preferredVersions: versions.filter(version => dep.alias && dep.pref ? semver.satisfies(version, dep.pref) : true),
+    preferredVersions: versions.filter(({ version }) => dep.alias && dep.pref ? semver.satisfies(version, dep.pref) : true),
   }
 }

--- a/patching/plugin-commands-patching/src/patchCommit.ts
+++ b/patching/plugin-commands-patching/src/patchCommit.ts
@@ -51,7 +51,7 @@ export async function handler (opts: install.InstallCommandOptions & Pick<Config
   await fs.promises.mkdir(patchesDir, { recursive: true })
   const patchedPkgManifest = await readPackageJsonFromDir(userDir)
   const pkgNameAndVersion = `${patchedPkgManifest.name}@${patchedPkgManifest.version}`
-  const gitTarballUrl = await getGitTarballUrlFormLockfile({
+  const gitTarballUrl = await getGitTarballUrlFromLockfile({
     alias: patchedPkgManifest.name,
     pref: patchedPkgManifest.version,
   }, {
@@ -183,10 +183,7 @@ async function areAllFilesInPkg (files: string[], basePath: string) {
   return equals(allFiles.sort(), files.sort())
 }
 
-async function getGitTarballUrlFormLockfile (dep: ParseWantedDependencyResult, opts: GetPatchedDependencyOptions) {
+async function getGitTarballUrlFromLockfile (dep: ParseWantedDependencyResult, opts: GetPatchedDependencyOptions) {
   const { preferredVersions } = await getVersionsFromLockfile(dep, opts)
-  if (preferredVersions[0]?.isGitHosted) {
-    return preferredVersions[0].tarball
-  }
-  return undefined
+  return preferredVersions[0]?.gitTarballUrl
 }

--- a/patching/plugin-commands-patching/src/patchCommit.ts
+++ b/patching/plugin-commands-patching/src/patchCommit.ts
@@ -14,8 +14,9 @@ import escapeStringRegexp from 'escape-string-regexp'
 import renderHelp from 'render-help'
 import tempy from 'tempy'
 import { writePackage } from './writePackage'
-import { parseWantedDependency } from '@pnpm/parse-wanted-dependency'
+import { type ParseWantedDependencyResult, parseWantedDependency } from '@pnpm/parse-wanted-dependency'
 import packlist from 'npm-packlist'
+import { type GetPatchedDependencyOptions, getVersionsFromLockfile } from './getPatchedDependency'
 
 export const rcOptionsTypes = cliOptionsTypes
 
@@ -50,9 +51,16 @@ export async function handler (opts: install.InstallCommandOptions & Pick<Config
   await fs.promises.mkdir(patchesDir, { recursive: true })
   const patchedPkgManifest = await readPackageJsonFromDir(userDir)
   const pkgNameAndVersion = `${patchedPkgManifest.name}@${patchedPkgManifest.version}`
+  const gitTarballUrl = await getGitTarballUrlFormLockfile({
+    alias: patchedPkgManifest.name,
+    pref: patchedPkgManifest.version,
+  }, {
+    lockfileDir,
+    modulesDir: opts.modulesDir,
+    virtualStoreDir: opts.virtualStoreDir,
+  })
   const srcDir = tempy.directory()
-  await writePackage(parseWantedDependency(pkgNameAndVersion), srcDir, opts)
-
+  await writePackage(parseWantedDependency(gitTarballUrl ? `${patchedPkgManifest.name}@${gitTarballUrl}` : pkgNameAndVersion), srcDir, opts)
   const patchedPkgDir = await preparePkgFilesForDiff(userDir)
   const patchContent = await diffFolders(srcDir, patchedPkgDir)
 
@@ -173,4 +181,12 @@ async function areAllFilesInPkg (files: string[], basePath: string) {
     cwd: basePath,
   })
   return equals(allFiles.sort(), files.sort())
+}
+
+async function getGitTarballUrlFormLockfile (dep: ParseWantedDependencyResult, opts: GetPatchedDependencyOptions) {
+  const { preferredVersions } = await getVersionsFromLockfile(dep, opts)
+  if (preferredVersions[0]?.isGitHosted) {
+    return preferredVersions[0].tarball
+  }
+  return undefined
 }

--- a/patching/plugin-commands-patching/test/patch.test.ts
+++ b/patching/plugin-commands-patching/test/patch.test.ts
@@ -366,11 +366,21 @@ describe('prompt to choose version', () => {
     prompt.mockResolvedValue({
       version: '5.3.0',
     })
-
+    prompt.mockClear()
     const output = await patch.handler(defaultPatchOption, ['chalk'])
 
-    expect(prompt.mock.calls[0][0].choices).toEqual(expect.arrayContaining(['5.3.0', '4.1.2']))
-    prompt.mockClear()
+    expect(prompt.mock.calls[0][0].choices).toEqual(expect.arrayContaining([
+      {
+        name: '4.1.2',
+        message: '4.1.2',
+        value: '4.1.2',
+      },
+      {
+        name: '5.3.0',
+        message: '5.3.0',
+        value: '5.3.0',
+      },
+    ]))
 
     const patchDir = getPatchDirFromPatchOutput(output)
     const tempDir = os.tmpdir()
@@ -506,6 +516,7 @@ describe('patch and commit in workspaces', () => {
         version: '1.0.0',
         dependencies: {
           'is-positive': '1.0.0',
+          hi: '0.0.0',
         },
       },
       {
@@ -514,6 +525,7 @@ describe('patch and commit in workspaces', () => {
         dependencies: {
           'is-positive': '1.0.0',
           'project-1': '1',
+          hi: 'github:zkochan/hi#4cdebec76b7b9d1f6e219e06c42d92a6b8ea60cd',
         },
       },
     ])
@@ -569,7 +581,6 @@ describe('patch and commit in workspaces', () => {
       workspaceDir: process.cwd(),
       saveLockfile: true,
       frozenLockfile: false,
-      fixLockfile: true,
     }, [patchDir])
 
     const { manifest } = await readProjectManifest(process.cwd())
@@ -645,6 +656,69 @@ describe('patch and commit in workspaces', () => {
     expect(fs.existsSync('./node_modules/is-positive/license')).toBe(false)
     expect(fs.readFileSync('../project-2/node_modules/is-positive/index.js', 'utf8')).not.toContain('// test patching')
     expect(fs.existsSync('../project-2/node_modules/is-positive/license')).toBe(true)
+  })
+
+  test('patch and patch-commit for git hosted dependency', async () => {
+    const { allProjects, allProjectsGraph, selectedProjectsGraph } = await readProjects(process.cwd(), [])
+    await install.handler({
+      ...DEFAULT_OPTS,
+      cacheDir,
+      storeDir,
+      allProjects,
+      allProjectsGraph,
+      dir: process.cwd(),
+      lockfileDir: process.cwd(),
+      selectedProjectsGraph,
+      workspaceDir: process.cwd(),
+      saveLockfile: true,
+    })
+
+    prompt.mockResolvedValue({
+      version: 'https://codeload.github.com/zkochan/hi/tar.gz/4cdebec76b7b9d1f6e219e06c42d92a6b8ea60cd',
+    })
+    prompt.mockClear()
+    const output = await patch.handler(defaultPatchOption, ['hi'])
+    expect(prompt.mock.calls[0][0].choices).toEqual(expect.arrayContaining([
+      {
+        name: '0.0.0',
+        message: '0.0.0',
+        value: '0.0.0',
+      },
+      {
+        name: '1.0.0',
+        message: '1.0.0',
+        value: 'https://codeload.github.com/zkochan/hi/tar.gz/4cdebec76b7b9d1f6e219e06c42d92a6b8ea60cd',
+        hint: 'Git Hosted',
+      },
+    ]))
+    const patchDir = getPatchDirFromPatchOutput(output)
+    expect(fs.existsSync(patchDir)).toBe(true)
+    expect(fs.readFileSync(path.join(patchDir, 'index.js'), 'utf8')).toContain('module.exports = \'Hi\'')
+    fs.appendFileSync(path.join(patchDir, 'index.js'), '// test patching', 'utf8')
+
+    await patchCommit.handler({
+      ...DEFAULT_OPTS,
+      allProjects,
+      allProjectsGraph,
+      selectedProjectsGraph,
+      dir: process.cwd(),
+      rootProjectManifestDir: process.cwd(),
+      cacheDir,
+      storeDir,
+      lockfileDir: process.cwd(),
+      workspaceDir: process.cwd(),
+      saveLockfile: true,
+      frozenLockfile: false,
+    }, [patchDir])
+
+    const { manifest } = await readProjectManifest(process.cwd())
+    expect(manifest.pnpm?.patchedDependencies).toStrictEqual({
+      'hi@1.0.0': 'patches/hi@1.0.0.patch',
+    })
+    const patchContent = fs.readFileSync('patches/hi@1.0.0.patch', 'utf8')
+    expect(patchContent).toContain('diff --git')
+    expect(patchContent).toContain('// test patching')
+    expect(fs.readFileSync('./project-2/node_modules/hi/index.js', 'utf8')).toContain('// test patching')
   })
 })
 

--- a/patching/plugin-commands-patching/tsconfig.json
+++ b/patching/plugin-commands-patching/tsconfig.json
@@ -25,6 +25,9 @@
       "path": "../../config/pick-registry-for-package"
     },
     {
+      "path": "../../fetching/pick-fetcher"
+    },
+    {
       "path": "../../lockfile/lockfile-file"
     },
     {

